Add document types and subtypes

Finemas opened the pull request:
Added model for document types and subtypes, update a document model for them and view at the last step.

Finemas committed:
Update domain and database models for document types

Finemas committed:
Add view for document types

Finemas committed:
Fix unable to write in text input field

Finemas committed:
Remove sorting document types

Finemas committed:
Update document types UI

Finemas committed:
TestFlight build

Finemas committed:
Change view model for new logic of doc types

Finemas committed:
Add selection for document types

Finemas committed:
Add view for subtypes

skorepak committed:
Show subtype selector if needed. Update validation. Auto-select single subtype

Finemas committed:
TestFlight build

Finemas committed:
Update date attribute

Finemas committed:
Add selected department to new document

Finemas committed:
Update document models

Removed unused notes and update attribute for storing subtypes

Finemas committed:
Fix saving document name

Finemas committed:
Fix twice loading documents

Finemas committed:
Add keyboard dismissing while name writing

skorepak commented on zScanner/Services/Network/Medicalc/Model/DocumentTypeNetworkModel.swift:
```diff
@@ -28,7 +28,8 @@ extension DocumentTypeNetworkModel {
             id: id,
-            name: display
+            name: display,
+            subtypes: subtypes?.map({ DocumentSubTypeDomainModel(id: $0.id, name: $0.display) }) ?? []
```
Use `subtypes?.map({ $0.toDomainModel })

The subtype should know how to map itself. Type should not be creating Instance of SubType

skorepak replied:
Should be same as in TypeDatabaseModel

Finemas replied:
It won't work because it can't convert DocumentTypeDomainModel to DocumentSubTypeDomainModel. Should I create a subtype network model?

jakubskorepapoq replied:
Yes :)

Finemas committed:
Move loading circle in department button

skorepak commented on zScanner/SupportFiles/Resources/cs.lproj/Localizable.strings:
```diff
@@ -67,7 +67,8 @@
 
 // MARK: - Form
-"form.listPicker.title" = "Typ dokumentu";
+"form.listPicker.docType.title" = "Typ dokumentu";
```
Use `documentType` and `documentSubType`, do not shorten the names.

skorepak commented on zScanner/ViewControllers/Documents/DocumentsListViewController.swift:
```diff
@@ -83,12 +83,12 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
                     
                 case .loading:
-                    self.rightBarButtons = [self.loadingItem]
+                    self.rightBarButtons = []
```
Where is the loading now indicated?

Finemas replied:
I moved it to the department button instead of.

skorepak commented on zScanner/Views/Departments/DepartmentButton.swift:
```diff
@@ -29,6 +29,12 @@ class DepartmentButton: PrimaryButton {
         didSet {
             backgroundColor = isSelected ? .primaryDark : .primary
+            
+            if isSelected {
```
Activity indicator have to be connected to the viewModel state, not to the selection of the button. 

We always need to react to the origin of the event, not to the side-effect. The side effect may change and the loading might break.

jakubskorepapoq commented on zScanner/ViewModels/Documents/DocumentsListViewModel.swift:
```diff
@@ -59,7 +57,7 @@ class DocumentsListViewModel {
     let disposeBag = DisposeBag()
     
-    private var lastSelectedDepartment: DepartmentDomainModel?
+    var lastSelectedDepartment: DepartmentDomainModel?
```
private(set)

jakubskorepapoq commented on zScanner/ViewControllers/Documents/DocumentsListViewController.swift:
```diff
@@ -83,12 +83,12 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
                     self.rightBarButtons = []
                     self.departmentsStackView.subviews.forEach({ ($0 as? UIButton)?.isSelected = false })
-                    self.coordinator.createNewDocument()
+                    self.coordinator.createNewDocument(with: self.documentsViewModel.lastSelectedDepartment!)
```
It's better to unwrap it. If you don't really have to, don't use force unwrapping. 
```
if let department = self.documentsViewModel.lastSelectedDepartment {
    self.coordinator.createNewDocument(with: department)
}
```

jakubskorepapoq commented on zScanner/ViewModels/Documents/DocumentsListViewModel.swift:
```diff
@@ -27,8 +27,6 @@ class DocumentsListViewModel {
         self.database = database
         self.networkManager = ikemNetworkManager
-        
-        loadDocuments()
```
Why did you remove it?

Finemas replied:
It was already called in view controller with updateDocuments method so it was called twice.

jakubskorepapoq replied:
I think there was some reason for it, but I don't remember it now. Let's see if it's gonna be ok :)

Finemas committed:
Update code format

jakubskorepapoq commented on zScanner/SupportFiles/Extensions/UIViewController.swift:
```diff
@@ -12,4 +12,16 @@ extension UIViewController {
+        view.addGestureRecognizer(tap)
+    }
+    
+    @objc func dismissKeyboard() {
```
you can use private here. this method should not be called form outside of this extension

jakubskorepapoq commented on zScanner/ViewControllers/Documents/DocumentsListViewController.swift:
```diff
@@ -83,12 +83,15 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
                     self.rightBarButtons = []
```
Please remove the `rightBarButtons` if they are not used anymore in the view controller

Finemas committed:
Add network model for document subtypes

Finemas committed:
Remove right bar buttons from documents list

Finemas committed:
Update selection of department button

The department should be deselected only after successfully download document types

Finemas committed:
Update view controller extension

Finemas committed:
Fix title of list picker

skorepak committed:
Fix button loading handling

skorepak committed:
Update string based on the UI
